// html-validate/dist/cjs/matchers-jestonly.js
'use strict';

var jestSnapshot = require('jest-snapshot');
var matchers = require('./matchers.js');
var matcherUtils = require('./matcher-utils.js');

function createMatcher$1() {
  function toMatchCodeframe(actual, ...rest) {
    const filename = this.testPath ?? "inline";
    const results = matchers.getResults(filename, actual);
    const snapshot = matcherUtils.codeframe(results).replaceAll(/\s+$/gm, "");
    return jestSnapshot.toMatchSnapshot.call(this, snapshot, ...rest);
  }
  return matcherUtils.diverge(toMatchCodeframe);
}

function toMatchInlineCodeframeImpl(context, actual, ...rest) {
  const filename = context.testPath ?? "inline";
  const results = matchers.getResults(filename, actual);
  const snapshot = matcherUtils.codeframe(results).replaceAll(/\s+$/gm, "");
  return jestSnapshot.toMatchInlineSnapshot.call(context, snapshot, ...rest);
}
function toMatchInlineCodeframe(actual, ...rest) {
  const context = {
    ...this,
    /* Capture the original stack frames as they are needed by "jest-snapshot"
     * to determine where to write the inline snapshots. When resolving the
     * promise the original stack frames are lost and the snapshot will be
     * written in this files instaed. */
    error: new Error("stacktrace")
  };
  if (matcherUtils.isThenable(actual)) {
    return actual.then((resolved) => toMatchInlineCodeframeImpl(context, resolved, ...rest));
  } else {
    return toMatchInlineCodeframeImpl(context, actual, ...rest);
  }
}
function createMatcher() {
  return toMatchInlineCodeframe;
}

exports.createMatcher = createMatcher;
exports.createMatcher$1 = createMatcher$1;
