// html-validate/dist/esm/matchers-jestonly.js
import { toMatchSnapshot, toMatchInlineSnapshot } from 'jest-snapshot';
import { g as getResults } from './matchers.js';
import { d as diverge, a as codeframe, i as isThenable } from './matcher-utils.js';

function createMatcher$1() {
  function toMatchCodeframe(actual, ...rest) {
    const filename = this.testPath ?? "inline";
    const results = getResults(filename, actual);
    const snapshot = codeframe(results).replaceAll(/\s+$/gm, "");
    return toMatchSnapshot.call(this, snapshot, ...rest);
  }
  return diverge(toMatchCodeframe);
}

function toMatchInlineCodeframeImpl(context, actual, ...rest) {
  const filename = context.testPath ?? "inline";
  const results = getResults(filename, actual);
  const snapshot = codeframe(results).replaceAll(/\s+$/gm, "");
  return toMatchInlineSnapshot.call(context, snapshot, ...rest);
}
function toMatchInlineCodeframe(actual, ...rest) {
  const context = {
    ...this,
    /* Capture the original stack frames as they are needed by "jest-snapshot"
     * to determine where to write the inline snapshots. When resolving the
     * promise the original stack frames are lost and the snapshot will be
     * written in this files instaed. */
    error: new Error("stacktrace")
  };
  if (isThenable(actual)) {
    return actual.then((resolved) => toMatchInlineCodeframeImpl(context, resolved, ...rest));
  } else {
    return toMatchInlineCodeframeImpl(context, actual, ...rest);
  }
}
function createMatcher() {
  return toMatchInlineCodeframe;
}

export { createMatcher$1 as a, createMatcher as c };
